Merge identical windows

ebd2 opened the pull request:
This optimizer is limited to merging WindowNodes with identical specifications in groups of WindowNodes in a plan that do not contain other types of PlanNodes.

In addition, I've added product tests covering a number of situations with mergeable windows. Note that window functions that take extra parameters result in a different plan than ones that do not. nth_value(), for example, takes a parameter specifying which value in the window to take.

These functions produce a plan with a ProjectNode that cannot be pruned by the PruneIdentityProjections optimizer, and so are special cases that should be tested.

@maciejgrzybek, @cawallin

ebd2 commented:
I'm also considering adding product tests where there are identical window specifcations, but where one of the functions produces the un-pruneable project node. Thoughts?

cawallin commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizersFactory.java:
```diff
@@ -67,7 +68,7 @@ public PlanOptimizersFactory(Metadata metadata, SqlParser sqlParser, FeaturesCon
         ImmutableList.Builder<PlanOptimizer> builder = ImmutableList.builder();
 
-        builder.add(new DesugaringOptimizer(metadata, sqlParser), // Clean up all the sugar in expressions, e.g. AtTimeZone, must be run before all the other optimizers
+        builder.add(new DesugaringOptimizer(metadata, sqlParser), // Clean up all the sugar in expressions, e.g. AtTimeZone, must be run bef,ore all the other optimizers
```
typo

cawallin commented on presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/windowDescriptionAABwithEP.sql:
```diff
@@ -0,0 +1,8 @@
+-- database: presto; groups: window;
+select
```
EP? (from the filename)

ebd2 replied:
Extra Parameters (or Extended Play if you prefer your MP3s in analog format stored on big black slices of vinyl). Open to a better name that isn't super long.

maciejgrzybek replied:
What about `twoAdjacentIdenticalWindowsFollowedByOneWithProjection`?

ebd2 replied:
Here and elsewhere, the projection node is an artifact of the current implementation of the window functionality. We shouldn't name the test after an implementation detail.

maciejgrzybek replied:
If circumstances change (e.g. projection node is not there anymore) we'll need to change tests anyway (because they might not be valid anymore).

ebd2 replied:
They may not be overly useful, but they wouldn't be rendered invalid. What's special about lead, lag, ntile, and nth_value is that they have additional parameters, not that those additional parameters end up on a projection node.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/WindowNode.java:
```diff
@@ -291,5 +291,28 @@ public Frame(
+                return false;
+            }
+            Frame other = (Frame) obj;
+            return Objects.equals(this.type, other.type) &&
```
What about `id` (from base class)?

maciejgrzybek commented on presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/windowDescriptionABAnoEP.sql:
```diff
@@ -0,0 +1,8 @@
+-- database: presto; groups: window;
```
`identicalWindowsSeparatedByNonIdentical`?

maciejgrzybek commented on presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/windowDescriptionABAwithEP.sql:
```diff
@@ -0,0 +1,8 @@
+-- database: presto; groups: window;
```
`twoIdenticalWindowsSeparatedByOneWithProjection` ?

maciejgrzybek commented on presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/windowDescriptionBAAwithEP.sql:
```diff
@@ -0,0 +1,8 @@
+-- database: presto; groups: window;
```
`twoIdenticalWindowsPrecededByNonIdentical`?

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,137 @@
+                    "Merge context is in an inconsistent state.");
+
+            ImmutableMap.Builder<Symbol, FunctionCall> functionsBuilder =
+                    functions.getOrDefault(key, new ImmutableMap.Builder<>());
```
`Builder` will be created even if not needed. I'd go for approach with checking if key exists in a map.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,137 @@
+            return signatures.put(key, signaturesBuilder) == null;
+        }
+
+        private class FunctionPair
```
`static`

cawallin commented:
also, you should probably write unit tests via this framework: https://github.com/prestodb/presto/pull/4404

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,137 @@
+            WindowNode.Specification key = window.getSpecification();
+
+            // Either both maps must contain the specification, or neither.
+            checkState(functions.containsKey(key) == signatures.containsKey(key),
```
Couldn't we merge both maps to avoid such a consistency checks?
E.g. Map<Specification, ImmutableMap.Builder<Symbol, Pair<FunctionCall, Signature>>>`or, instead of`Pair`you can use some specialized, more descriptive structure?
You even have it already -`FunctionPair`.

kokosing replied:
I would add to @maciejgrzybek  comment that having a map of map builders is to me overly complicated. I would extract some entity and simplify these maps. Lets name this entity `SymbolMapping` then I would have a single `Map<WindowNode.Specyfication, SymbolMapping> mapping`.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizersFactory.java:
```diff
@@ -68,7 +68,7 @@ public PlanOptimizersFactory(Metadata metadata, SqlParser sqlParser, FeaturesCon
     {
```
Please make sure to squash it with dab0092ea5fbc5b1344200651100e10d53f0eeda.

maciejgrzybek commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -30,6 +30,38 @@
 
 import static com.google.common.base.Preconditions.checkState;
 
+/**
```
Shouldn't we provide a unit tests describing this behavior instead of comment?

maciejgrzybek commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,108 @@
+
+    public TestMergeIdenticalWindows()
+    {
+        this.queryRunner = new LocalQueryRunner(testSessionBuilder()
```
Shouldn't this be created per unit-test in `@setup` method?

ebd2 replied:
https://github.com/prestodb/presto/pull/4404/files#diff-42e96520332fe6de8a8ff605e9f8db12R44

I guess it depends how stateful it is, but @kokosing's initial test suite creates it in the constructor.

maciejgrzybek replied:
I'm only concerned about session. Maybe it does not matter though.

kokosing replied:
Creating `queryRunner` is expensive. It shouldn't contain any state, unless you do `DELETE` or `INSERT` ;)

ebd2 replied:
Thanks for fielding that one :-)

maciejgrzybek commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,108 @@
+                                                any())))));
+    }
+
+    private FunctionCall fc(String name, String... args) {
```
`static`

maciejgrzybek commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,108 @@
+                .setExperimentalSyntaxEnabled(true)
+                .setDistributedIndexJoinsEnabled(false)
+                .setOptimizeHashGeneration(true);
+        Provider<List<PlanOptimizer>> optimizerProvider = () -> new ImmutableList.Builder<PlanOptimizer>().add(
```
Why builder instead of `ImmutableList.of(...)`?

maciejgrzybek commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,108 @@
+    {
+        FeaturesConfig featuresConfig = new FeaturesConfig()
+                .setExperimentalSyntaxEnabled(true)
+                .setDistributedIndexJoinsEnabled(false)
```
Is this meaningful for this rewriter? If not maybe skip it or at least put comment next to this line?

ebd2 commented:
This is wildly ugly and needs some cleanup. I plan to do that and add more unit tests next week.

In the meantime, @kokosing: I'd appreciate a look from you on my use of the DSL and addition to it. Thanks!

ebd2 commented:
I've cleaned things up and added unit tests. I've also renamed the product tests to be consistent with the names of the unit tests.

Having written the unit tests, I'm in favor of retaining the block comment at the top of MergeIdenticalWindows, as it shows the plan pre and post optimizer, whereas the unit test only shows the post-optimization state.

I'd particularly appreciate thoughts on the DSL changes, particularly the FunctionCall and AnyQualifiedNameReference stuff.

@maciejgrzybek, @cawallin, @kokosing

kokosing commented on presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/windowSpecificationAABcp.sql:
```diff
@@ -0,0 +1,8 @@
+-- database: presto; groups: window;
```
I would merge `sql` and `result` files together. It will limit convention tests files and it is easier to read when you have a query and results in one place.

Moreover, (optionally - up to you) I would go one step further and merge all your window tests into one file. Here is a syntax of such convention tests:

```
-- database: hive; groups: example_smoketest,group2
-- delimiter: |; ignoreOrder: false; types: INTEGER|VARCHAR|INTEGER|VARCHAR
--! name: query_1
SELECT * FROM nation WHERE id=0
--!
0|ALGERIA|0| haggle. carefully final deposits detect slyly agai|
--! name: query_2
-- groups: additional_group
SELECT * FROM nation WHERE id=1
1|USA|1| foo bar|
--!
```

kokosing replied:
The test name `WindowSpecificationAABcp` does tell me nothing. Please add a comment above the query with explanation of test purpose.

kokosing replied:
Also consider adding some tests to `AbstractTestQueries` class. This class is run as unit tests and it is much easier to debug. I would leave only one test as a product test and move rest of them to that class (this kind of approach is called a [TestPyramid](http://martinfowler.com/bliki/TestPyramid.html).

ebd2 replied:
Comments added.
I propose merging the .sql and .results files together for all of the convention-based tests in here in a separate change. Thoughts?

kokosing replied:
sounds ok

kokosing commented on presto-product-tests/src/main/resources/sql-tests/testcases/window_functions/windowSpecificationBcpAA.sql:
```diff
@@ -0,0 +1,8 @@
+suppkey, orderkey, partkey,
+nth_value(quantity, 4) over (partition by partkey order by orderkey rows between UNBOUNDED preceding and CURRENT ROW) nth_value_quantity_B,
+sum(quantity) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_quantity_A,
+sum(discount) over (partition by suppkey order by orderkey rows between unbounded preceding and current row) sum_discount_A
```
Does window function can be only placed in select list of the query? Can you put a window function into an ORDER BY (HAVING or GROUP BY) clause? Will such window merge together?

ebd2 replied:
ORDER BY looks good - it produces the same basic unoptimized plan as when you don't include ORDER BY

GROUP BY produces the error `GROUP BY clause cannot contain aggregations or window functions: [...`

HAVING generates an error: `sum(double):double is not a scalar function`, which seems like it might be a cryptic error meaning the same thing?

The postgresql [docs](https://www.postgresql.org/docs/current/static/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS) indicate that window functions can only be used in SELECT and ORDER BY.

The 2011 spec says "The &lt;search condition> shall not contain a &lt;window function> without an intervening &lt;query expression>." I'll try to come up with a conforming query and see what happens :-)

ebd2 replied:
I also need to try out what happens with a window function in a CTE.

ebd2 replied:
The CTE case is straightforward for a single CTE: The WindowNodes in the CTE show up basically as they do in a normal query.

If you have 2 CTEs and join them, we do not merge functions from WindowNodes with identical specifications across the left and right sources of the join. Without digging deeper, this seems likely to be the correct behavior. The reason this happens is that the symbol for the result of the function is different. If it were to be the same we'd merge. As I mentioned, I'm skeptical about this absent some quality time with some smart people around a whiteboard.

If we're OK counting on the symbols on the left and right sides of the join being different, we can leave MIW as-is. If not, we need to create new MIW.Context for each side of the join.

Thoughts? @kokosing, @cawallin, @maciejgrzybek

ebd2 replied:
Ugh, right reason, wrong symbol. It's the symbols in the WindowNode.specification that have different names, not the ones in the WindowNode.functionHandles/WindowNode.windowFunctions.

The same question applies about depending on this vs making it explicit that the sides of a join need to have separate contexts.

kokosing replied:
IMO it is cleaner to have separate contexts for each side of join.

ebd2 replied:
In offline discussion @maciejgrzybek and @mattsfuller thought the different symbols were sufficient and keep the code more generic. I'll leave it as-is for the moment, but add a test covering joining 2 CTEs with WindowNodes that have identical specs.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,174 @@
+ *    `--WindowNode(Specification: A, Functions: [sum(something)])
+ *       `--WindowNode(Specification: B, Functions: [sum(something)])
+ *          `-- ProjectNode(...)
+ *             `--WindowNode(Specification: A, Functions: [avg(something)])
```
Is there any reason (algebraic) why such windows are not merged? Or is just not yet supported. It sounds like big limitation. If it is possible I would strive to support this as it seems important. ProjectNode can appear everywhere. I think it could be possible by pushing down ProjectNode under WindowNodes (or by pulling it up).

ebd2 replied:
This is a first-step approach to solve the problem where there's a query with a lot of window functions in the select part of the query, which is a problem we have somebody actually asking us to solve.

What this approach doesn't solve for that case is when we have window functions that take constant parameters in that big list, because those introduce ProjectNodes into the chains of WindowNodes.

I did some experimentation merging the window functions into the WindowNodes both above and below the ProjectNodes, but that introduces the requirement that we change the projection list in the ProjectNode. That seems heavy-handed for this initial work.

I'm planning on the next step being to look at whether the correct approach to solving the problem of window functions that take constant parameters is to try to eliminate the ProjectNode in the plan, or whether it's to do as suggested and solve the general problem. 

That'll require a closer look at whether there are cases where we can't merge across non-WindowNodes for algebraic reasons and handling those cases appropriately.

kokosing replied:
thanks for explanation

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,174 @@
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan, new Context());
+    }
+
+    private static class Context
```
I think that `Context` class is less important than `Rewriter` so I would put it underneath.

ebd2 replied:
Done.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,174 @@
+            boolean retainNode = mergeContext.collectFunctions(node);
+
+            // Don't merge window functions across nodes that aren't WindowNodes
+            if (node.getClass().isInstance(node.getSource())) {
```
explicitly compare it with `WindowNode` instead of node.getClass

ebd2 replied:
Why? That specifies the type in two places (parameter list and here) rather than just one (parameter list). OaOO?

kokosing replied:
I understand. However, with my approach the code is more readable. When I see code like `node.getClass()` I think that I am in generic handling node function, while here it is opposite - very specific WindowNode function.

Up to you.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,174 @@
+            WindowNode.Specification key = window.getSpecification();
+
+            ImmutableMap.Builder<Symbol, FunctionCall> functionBuilder = functions.remove(key);
+            ImmutableMap.Builder<Symbol, Signature> signatureBuilder = signatures.remove(key);
```
why do you remove them? Maybe instead of removing you can not add if a function or signature already exists in collection. In such case you collect all information at the end of `Rewrite`, IMO thanks to that it will be easier to debug.

ebd2 replied:
I'm not sure I understand what you're proposing with not adding functions or signatures that already exist. Can you please clarify?

kokosing replied:
No need for explanation. You already did what I wanted ;)

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java:
```diff
@@ -164,4 +187,31 @@ private String indentString(int indent)
+        @Override
+        public int hashCode()
+        {
+            throw new UnsupportedOperationException();
```
Can you please explain?

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java:
```diff
@@ -164,4 +187,31 @@ private String indentString(int indent)
         return Strings.repeat("    ", indent);
     }
+
+    private static class AnyQualifiedNameReference extends QualifiedNameReference
```
It sounds like a hack to extend `QualifiedNameReference`. Maybe you can just extend from `Expression` and then you could name this as `AnyReference`

ebd2 replied:
I've changed it to AnyExpression, since function arguments are expressions, not necessarily references.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/WindowMatcher.java:
```diff
@@ -0,0 +1,83 @@
+            }
+            return false;
+        }
+        return true;
```
instead of using `goto` you could check if the list of `expectedCalls` is empty at the end of this function

ebd2 replied:
I think we need both. See the comments in the new one and let me know if you agree.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,150 @@
+        });
+    }
+
+    private Plan unitPlan(@Language("SQL") String sql)
```
I would avoid to run only selected list of optimizers. All the optimizers have to work with each other. Simply speaking I would just put your tests into `TestLogicalPlanner` and use its `assertPlan` method.

ebd2 replied:
Oof. This was the subject of a lot of discussion at the Boston office with @maciejgrzybek and @cawallin.

The short version is that if you run all the optimizers you aren't really running a true unit test, and if you get a failure, you don't necessarily know whether it's because MIW is broken or because some other optimizer changed the plan.

With the two types of tests in here we can separately answer the questions of:
1. Is MIW working as it's supposed to be? The tests running the minimal set of optimizers can tell us this.
2. Has some other optimizer changed the plan? The test running the full set of optimizers answers this, though it isn't actually meaningful unless we know the answer to question 1 is "yes".

The tests that use the minimal set of optimizers are closer to true "unit" tests in that they verify the behavior of MIW with as few external dependencies as possible. Unfortunately, those dependencies include the parser and analyzer, hence "unit" in quotes.

The conclusion of the discussion here was that nobody really likes the fact that this isn't a true unit test, but that we aren't going to try to generalize a solution from a sample size of one.

maciejgrzybek replied:
Small thing to add, we chose to use this approach instead of "true" unit test because it's much more readable with SQL queries instead of building optimizer Nodes manually.

ebd2 replied:
It's also dramatically less typing, and a lot more self-maintaining than trying to build the PlanNodes manually...

kokosing replied:
I see your point. Nevertheless, I think you still need to have a test which uses the same planner configuration as it used in production to avoid "false positives" (when the test passes, but functionality does not work on production). It would be sad having TMIW passing and this rewrite was not being used in production. This kind of bug is hard to find as this is an optimization rewrite, query passes anyway (with or without the rewrite). Product tests won't catch it for sure.

From the other side, the chance that other rule would break the result of MIW, or that the MIW is not used seems to be low. I am not going argue about this comment. I shared my concern, but it is you who makes the decision.

ebd2 replied:
Is this the type of test you're looking for?
https://github.com/Teradata/presto/pull/222/files#diff-d10d6da7ffa24365986581d2ea2f12d6R57

This is the one that answers the question of whether or not somebody has reordered the optimizers in such a way that makes the optimizer ineffective (i.e. question 2 above). If so, do I need to call it out better in a comment?

kokosing replied:
The link you posted points to the commit where there are only "unit" tests. 

 Exactly I meant the test which answers for question 2 from the above.

ebd2 replied:
Cleared up on offline discussion. The linked test is different from the other ones. I'll add a comment calling it out and summarizing why we have 2 different kinds of tests.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,150 @@
+                        window(ImmutableList.of(functionCall("sum", "discount"), functionCall("lag", "quantity", "*", "*")),
+                                project(
+                                        window(ImmutableList.of(functionCall("sum", "quantity")),
+                                                any())))));
```
assert that child of window is not a window, to make sure that window node was pruned.

ebd2 replied:
Fixed. Nice idea!

ebd2 replied:
Hmmm. This one is weird, it's kind of covered implicitly by the project node. Should I try to add an anyNot(WindowNode.class) above the first window or is that being overly paranoid?

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestMergeIdenticalWindows.java:
```diff
@@ -0,0 +1,150 @@
+        assertUnitPlan(sql,
+                anyTree(
+                        window(ImmutableList.of(functionCall("sum", "discount"), functionCall("lag", "quantity", "*", "*")),
+                                project(
```
How this `project` get in? You wrote in `MergeIdenticalWindows` that this optimizer works when all the windows are grouped in the plan (without any nodes inside).

ebd2 replied:
The project node is an artifact of how lead, lag, ntile, and nth_value are implemented. Their constant parameters are added as expressions in a project node that is the source for the WindowNode that contains the function.

select lag(value, 1, 0.0) from ...
results in a plan that looks something like this

```
OutputNode
`-- ...
   `--WindowNode(Specification: $somespec, Functions: [lead(value, expr_4, expr_8)])
      `--ProjectNode(... value, expr_4 := 1, expr_8 := 0.0 ...)
         `-- ...
```

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeIdenticalWindows.java:
```diff
@@ -0,0 +1,154 @@
+
+            private FunctionPair()
+            {
+                this.windowFunctions = new ImmutableMap.Builder<>();
```
initialize this fields at declaration, then you could remove this constructor

ebd2 replied:
done.

ebd2 commented:
Closing to open FB PR.
